src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -61,63 +60,90 @@ internal override BoundStatement BindUsingStatementParts(DiagnosticBag diagnosti
-            }
-            else
-            {
-                ImmutableArray<BoundLocalDeclaration> declarations;
-                originalBinder.BindForOrUsingOrFixedDeclarations(declarationSyntax, LocalDeclarationKind.UsingVariable, diagnostics, out declarations);
+                else
+                {
+                    ImmutableArray<BoundLocalDeclaration> declarations;
+                    originalBinder.BindForOrUsingOrFixedDeclarations(declarationSyntax, LocalDeclarationKind.UsingVariable, diagnostics, out declarations);
 
-                Debug.Assert(!declarations.IsEmpty);
+                    Debug.Assert(!declarations.IsEmpty);
 
-                declarationsOpt = new BoundMultipleLocalDeclarations(declarationSyntax, declarations);
+                    declarationsOpt = new BoundMultipleLocalDeclarations(declarationSyntax, declarations);
 
-                TypeSymbol declType = declarations[0].DeclaredType.Type;
+                    TypeSymbol declType = declarations[0].DeclaredType.Type;
 
-                if (declType.IsDynamic())
-                {
-                    iDisposableConversion = Conversion.ImplicitDynamic;
-                }
-                else
-                {
-                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                    iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declType, iDisposable, ref useSiteDiagnostics);
-                    diagnostics.Add(declarationSyntax, useSiteDiagnostics);
-
-                    if (!iDisposableConversion.IsImplicit)
+                    if (declType.IsDynamic())
                     {
-                        if (!declType.IsErrorType())
+                        iDisposableConversion = Conversion.ImplicitDynamic;
+                    }
+                    else
+                    {
+                        HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                        iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declType, iDisposable, ref useSiteDiagnostics);
+                        diagnostics.Add(declarationSyntax, useSiteDiagnostics);
+
+                        if (!iDisposableConversion.IsImplicit)
                         {
-                            Error(diagnostics, ErrorCode.ERR_NoConvToIDisp, declarationSyntax, declType);
+                            if (!declType.IsErrorType())
+                            {
+                                Error(diagnostics, hasAwait ? ErrorCode.ERR_NoConvToIAsyncDisp : ErrorCode.ERR_NoConvToIDisp, declarationSyntax, declType);
+                            }
+
+                            hasErrors = true;
                         }
+                    }
+                }
 
-                        hasErrors = true;
+                if (!hasErrors && hasAwait)
+                {
+                    TypeSymbol taskType = this.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task);
```

Comment:
>GetWellKnownType [](start = 59, length = 16)

I think Binder has a helper that gets the type and reports use-site diagnostics.

Reply:
They don't accept a `SyntaxToken` or a `Location`. I don't want to multiple overloads, so I just added one to `ReportUseSiteDiagnostics`.

---
In reply to: [160563783](https://github.com/dotnet/roslyn/pull/23961#discussion_r160563783) [](ancestors = 160563783)